File: 卡码/代码随想录/JS/二叉树/03-二叉树修改与构造/LeetCode_0105.js
//105. 从前序与中序遍历序列构造二叉树
import TreeNode from "../TreeNode.js";

const traversal=function(preOrder,preLeft,preRight,inOrder,inLeft,inRight){
  if(inLeft>inRight) return null;
  const root=new TreeNode(preOrder[preLeft],null,null);
  if(inLeft==inRight) return root;
  let rootIndex=0;
  for(let i=inLeft;i<=inRight;i++){
    if(inOrder[i]==root.val){
      rootIndex=i;
      break;
    }
  }
  const leftLength=rootIndex-inLeft;
  root.left=traversal(preOrder,preLeft+1,preLeft+leftLength,inOrder,inLeft,rootIndex-1);
  root.right=traversal(preOrder,preLeft+leftLength+1,preRight,inOrder,rootIndex+1,inRight);
  return root;
}

const buildTree = function(preorder, inorder) {
  if(preorder.length==0||inorder.length==0) return null;
  return traversal(preorder,0,preorder.length-1,inorder,0,inorder.length-1);
};

File: 卡码/代码随想录/1/06-二叉树/Java/binaryTree/src/main/java/com/kurumi/binarytree/_03_二叉树修改与构造/LeetCode_0105.java
//105. 从前序与中序遍历序列构造二叉树
package com.kurumi.binarytree._03_二叉树修改与构造;

import com.kurumi.binarytree.TreeNode;

public class LeetCode_0105 {

}
class Solution {
  public TreeNode traversal(int[] inOrder, int inLeft, int inRight, int[] preOrder, int preLeft, int preRight) {
    if (inRight < inLeft) return null;
    TreeNode root = new TreeNode(preOrder[preLeft]);
    if (inRight == inLeft) {
      return root;
    }
    int rootIndex = 0;
    for (int i = inLeft; i <= inRight; i++) {
      if (inOrder[i] == root.val) {
        rootIndex = i;
        break;
      }
    }
    int leftLength = rootIndex - inLeft;
    root.left = traversal(inOrder, inLeft, rootIndex - 1, preOrder, preLeft+1, preLeft + leftLength);
    root.right = traversal(inOrder, rootIndex + 1, inRight, preOrder, preLeft+1+leftLength, preRight);
    return root;
  }
  public TreeNode buildTree(int[] preorder, int[] inorder) {
    if(preorder.length==0||inorder.length==0) return null;
    return traversal(inorder, 0, inorder.length-1, preorder, 0, preorder.length-1);
  }
}
